beefy-merkle-root/src/lib.rs

```diff
@@ -0,0 +1,807 @@
+where
+	H: Hasher,
+	V: Visitor,
+	I: Iterator<Item = Hash>,
+{
+	let upper = Vec::with_capacity(leaves.size_hint().0);
+	let mut next = match merkelize_row::<H, _, _>(leaves, upper, visitor) {
+		Ok(root) => return root,
+		Err(next) if next.is_empty() => return Hash::default(),
+		Err(next) => next,
+	};
+
+	let mut upper = Vec::with_capacity((next.len() + 1) / 2);
+	loop {
+		visitor.move_up();
+
+		match merkelize_row::<H, _, _>(next.drain(..), upper, visitor) {
+			Ok(root) => return root,
+			Err(t) => {
+				// swap collections to avoid allocations
+				upper = next;
+				next = t;
+			}
+		};
+	}
+}
+
+/// A generated merkle proof.
+///
+/// The structure contains all necessary data to later on verify the proof and the leaf itself.
+#[derive(Debug, PartialEq, Eq)]
+pub struct MerkleProof<T> {
+	/// Root hash of generated merkle tree.
+	pub root: Hash,
+	/// Proof items (does not contain the leaf hash, nor the root obviously).
+	///
+	/// This vec contains all inner node hashes necessary to reconstruct the root hash given the
+	/// leaf hash.
+	pub proof: Vec<Hash>,
+	/// Number of leaves in the original tree.
+	///
+	/// This is needed to detect a case where we have an odd number of leaves that "get promoted"
+	/// to upper layers.
+	pub number_of_leaves: usize,
+	/// Index of the leaf the proof is for.
+	pub leaf_index: usize,
+	/// Leaf content.
+	pub leaf: T,
+}
+
+/// A trait of object inspecting merkle root creation.
+///
+/// It can be passed to [`merkelize_row`] or [`merkelize`] functions and will be notified
+/// about tree traversal.
+trait Visitor {
+	/// We are moving one level up in the tree.
+	fn move_up(&mut self);
+
+	/// We are creating an inner node from given `left` and `right` nodes.
+	///
+	/// Note that in case of last odd node in the row `right` might be empty.
+	/// The method will also visit the `root` hash (level 0).
+	///
+	/// The `index` is an index of `left` item.
+	fn visit(&mut self, index: usize, left: &Option<Hash>, right: &Option<Hash>);
+}
+
+/// No-op implementation of the visitor.
+impl Visitor for () {
+	fn move_up(&mut self) {}
+	fn visit(&mut self, _index: usize, _left: &Option<Hash>, _right: &Option<Hash>) {}
+}
+
+/// Construct a Merkle Proof for leaves given by indices.
+///
+/// The function constructs a (partial) Merkle Tree first and stores all elements required
+/// to prove requested item (leaf) given the root hash.
+///
+/// Both the Proof and the Root Hash is returned.
+///
+/// # Panic
+///
+/// The function will panic if given [`leaf_index`] is greater than the number of leaves.
+pub fn merkle_proof<H, I, T>(leaves: I, leaf_index: usize) -> MerkleProof<T>
+where
+	H: Hasher,
+	I: IntoIterator<Item = T>,
+	I::IntoIter: ExactSizeIterator,
+	T: AsRef<[u8]>,
+{
+	let mut leaf = None;
+	let iter = leaves.into_iter().enumerate().map(|(idx, l)| {
+		let hash = H::hash(l.as_ref());
+		if idx == leaf_index {
+			leaf = Some(l);
+		}
+		hash
+	});
+
+	struct ProofCollection {
```

Comment:
I assume this is the path for a single proof, i.e. all the hashes from leaf `position` to the root. This is not a collection of multiple proofs?

Reply:
Yes, we only support single-leaf proof here.